### ISSUE TITLE
fixing subsets of variables

### ISSUE THREAD

**Issue by jameshensman** [GPflow/GPflow#305]
Joe Mellor asks:

I have been playing about with it [GPflow] this week and was interested in the
ability to, for example, hold some of the variances in an ARD-enabled
kernel fixed, while allowing to change others.
My understanding is that this is not possible using the fixed flag on
Param of interest, since the Param is associated with the entire array of
variances and not for an individual value in the array.
I thought I might be able to roll my own kernel by essentially replacing a
single Param object containing the variances by a list of Param objects
containing the variance for each dimension. The fixed flag could then be
toggled for each Param object in the list.
I wondered if you had thought about this at all? and perhaps had a more
elegant solution?

**Comment by jameshensman** [comment]
You're right, it's not possible to fix (or otherwise constrain) subsets of variables in GPflow. This is one functionality that we removed from the GPy framework, because it's not a very common use case, and introduces significant complexity at the core of GPflow. 

You're correct in thinking that you can roll your own kernel in the way that you've suggested. That should work just fine. 

An alternative that might be of broader interest is to roll your own transformation. I'll have a think about what that might look like, and get back to you.

**Comment by jameshensman** [comment]
A transform along the lines of the following should do the trick, though this is untested. 

To keep positive constraints also, you'll need to have another transform applied within this one. See DiagMatrix (in GPflow.transforms) for example.

```python
class FixPartial(GPflow.transforms.Transform):
    def __init__(self, fixed_inds, fixed_vals):
        GPflow.transforms.Transform.__init__(self)
        self.fixed_inds, self.fixed_vals = fixed_inds, fixed_vals

    def forward(self, x):
        total_size = x.size + self.fixed_vals.size
        nonfixed_inds = np.setdiff_1d(np.arange(total_size), self.fixed_inds)
        y = np.empty(total_size)
        y[nonfixed_inds] = x
        y[self.fixed_inds] = self.fixed_vals
        return y

    def backward(self, y):
        nonfixed_inds = np.setdiff_1d(np.arange(y.size), self.fixed_inds)
        return y.flatten()[nonfixed_inds]

    def tf_forward(self, x):
        total_size = x.size + self.fixed_vals.size
        nonfixed_inds = np.setdiff_1d(np.arange(total_size), self.fixed_inds)
        return tf.dynamic_stitch([self.fixed_inds, nonfixed_inds], [self.fixed_vals, x])

    def tf_log_jacobian(self, x):
        return 0.0

    def __str__(self):
        return 'PartiallyFixed'

    def free_state_size(self, variable_shape):
        total_size = np.prod(variable_shape)
        return total_size - self.fixed_vals.size
```

You could apply this transform to a variable by doing
```python
k = GPflow.kernels.Matern32(input_dim=4, ARD=True)
fixed_inds = np.array([0, 1])
fixed_vals = np.array([2.3, 4.5])
k.lengthscales.transform = FixPartial(fixed_inds, fixed_vals)
```

**Comment by markvdw** [comment]
I think this would be a great feature, with an implementation that doesn't increase complexity much!

**Comment by Bonnevie** [comment]
I'm thinking this scheme carries over to other "partial constraints"? Could this scheme be generalized so that you give it a transform and a fixed index set and then it returns a new partial transform? Sort of like a "class decorator".

**Comment by Bonnevie** [comment]
Haven't tested it completely, but the following seems to allow application of a positivity transform to a particular column, relying on calls to the full positivity transform instead of re-implementation of the functionality. I can't see why it wouldn't generalize to other transformations, as long as the same size is maintained (which I'll admit isn't a given). 
 
```
class PositiveColumn(GPflow.transforms.Transform):
    def __init__(self, column):
        GPflow.transforms.Transform.__init__(self)
        self.column = column
        self.positive = GPflow.transforms.positive
        
    def forward(self, x):
        nc = x.shape[1]
        y = np.zeros_like(x)
        other_columns = np.setdiff1d(np.arange(nc), self.column)
        y[:,other_columns] = x[:, other_columns]
        y[:, self.column] = self.positive.forward(x[:, self.column])
        return y

    def backward(self, y):
        nc = y.shape[1]
        x = np.zeros_like(x)
        other_columns = np.setdiff1d(np.arange(nc), self.column)
        x[:,other_columns] = y[:, other_columns]
        x[:, self.column] = self.positive.backward(y[:, self.column])
        return x

    def tf_forward(self, x):
        nc = x.shape[1]
        other_columns = np.setdiff1d(np.arange(nc), self.column)
        all_columns = tf.unstack(x, axis=1)
        all_columns[self.column] = self.positive.tf_forward(all_columns[self.column])
        return tf.stack(all_columns, axis = 1)
    
    def tf_log_jacobian(self, x):
        return -tf.reduce_sum(tf.log(1. + tf.exp(-tf.slice(x, [0, self.column], [-1, 1]))))

    def __str__(self):
        return 'PositiveColumn'
```

**Comment by markvdw** [comment]
Ah so you want to be able to put different transformations on different parts of a Param? So you want a single column to be constrained to positive, while all others are free?

What's your use case for this? Could you not just define different parameters?

**Comment by Bonnevie** [comment]
I also just realized that the `x` and `y` being fed to `.transform` are always flattened vectors, sigh. 

I need to have a SVGP model with `num_latents=2` where there's an extra likelihood term, which I have managed to introduce fairly painlessly. But now I run into the issue that for the model to optimize stably, I need the `q_mu` of one of the two latents to be positive-constrained. 

I was just hoping to be able to build on top of SVGP instead of having to start over from the beginning.

**Comment by markvdw** [comment]
If I were you, I would define another `q_mu`, and then concatenate the two to get the `q_mu`s used in the rest of `SGPR`. E.g. define `q_mu_free` and `q_mu_pos` as `Param`s. Then make a property `q_mu` that returns a tf Tensor of the two concatenated in the right way. This would allow you to specify separate transforms for each, without changing a thing in the rest of the code.

**Comment by Bonnevie** [comment]
that's probably pretty sensible, I'll try that.
